### PR TITLE
Assignment 4-Miguel Ángel Suárez

### PR DESCRIPTION
**Pull request by miguelangel-suarez**
Assignment 4 added to the corresponding directory from Miguel Ángel Suárez Ucieda

### CODE REVIEW

File: Assignment4/miguelangel-suarez-22c010/task06.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+g.add((ns.JaneSmith, VCARD.Given, Literal("Jane")))
+g.add((ns.JaneSmith, VCARD.Family, Literal("Smith")))
+g.add((ns.JaneSmith, VCARD.FN, Literal("Jane Smith")))
+for s, p, o in g:
+  print(s,p,o)
+
+"""**TASK 6.5: Add UPM as the university where John Smith works**"""
+
+g.add((ns.UPM, RDF.type, ns.University))
+g.add((VCARD.Works_at, RDF.type, RDFS.Datatype))
```
Comment (dgarijo) [inline Assignment4/miguelangel-suarez-22c010/task06.py:66]: This is a bad practice: you should not redefine the vocabulary created by another organization. It's like creating your own google.com URI. Instead, define a new property on a new namespace

File: Assignment4/miguelangel-suarez-22c010/task07.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+from rdflib.plugins.sparql import prepareQuery
+ns = Namespace("http://somewhere#")
+vcard = Namespace("http://www.w3.org/2001/vcard-rdf/3.0#")
+"""With RDFLib"""
+for s,p,o in g.triples((None, RDFS.subClassOf, ns.LivingThing)):
+  print(s)
+
+"""With SPARQL"""
+q1 = prepareQuery("SELECT ?x WHERE {?x rdfs:subClassOf ns:LivingThing}", initNs = {"rdfs":RDFS, "ns":ns})
```
Comment (dgarijo) [inline Assignment4/miguelangel-suarez-22c010/task07.py:36]: for both sparql and rdflib: you have not added the means to list subclasses of subclasses of living thing. For rdflib, use a recursive function. For sparql, use `*`

File: Assignment4/miguelangel-suarez-22c010/task07.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+"""**TASK 7.2: List all individuals of "Person" with RDFLib and SPARQL (remember the subClasses)**
+
+"""
+
+"""With RDFLib"""
+for persona, p, o in g.triples((None, RDF.type, ns.Person)):
+  print(persona)
+for subclase, p, o in g.triples((None,RDFS.subClassOf,ns.Person)):
+  for persona, p, o in g.triples((None,RDF.type,subclase)):
+    print(persona)
```
Comment (dgarijo) [inline Assignment4/miguelangel-suarez-22c010/task07.py:50]: same as above

File: Assignment4/miguelangel-suarez-22c010/task07.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+for r in g.query(q1):
+  print(r)
+
+"""**TASK 7.3: List all individuals of "Person" or "Animal" and all their properties including their class with RDFLib and SPARQL. You do not need to list the individuals of the subclasses of person**
+
+"""
+
+"""With RDFLib"""
+for elem,p,o in g.triples((None, RDF.type, ns.Animal)):
+  for s,t,u in g.triples((elem,None,None)):
```
Comment (dgarijo) [inline Assignment4/miguelangel-suarez-22c010/task07.py:64]: what about Person?

File: Assignment4/miguelangel-suarez-22c010/task07.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    print(s,t,u)
+
+"""With SPARQL"""
+q1 = prepareQuery('''
+    SELECT ?x ?relacion
+    WHERE{
+            ?x rdf:type ns:Animal.
+            ?x ?relacion ?a
+    }''', initNs = {"rdf": RDF, "ns": ns}
```
Comment (dgarijo) [inline Assignment4/miguelangel-suarez-22c010/task07.py:73]: missing Person

File: Assignment4/miguelangel-suarez-22c010/task07.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+for r in g.query(q1):
+  print(r)
+
+"""**TASK 7.4:  List the name of the persons who know Rocky**"""
+
+"""With RDFLib"""
+FOAF = Namespace("http://xmlns.com/foaf/0.1/")
+for persona,p,o in g.triples((None, RDF.type, ns.Person)):
+  for a,q,r in g.triples((None, FOAF.knows, persona)):
+    print(a)
```
Comment (dgarijo) [inline Assignment4/miguelangel-suarez-22c010/task07.py:84]: You are listing all the entities that know a Person, not what the exercise says

File: Assignment4/miguelangel-suarez-22c010/task07.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+"""With SPARQL"""
+q1 = prepareQuery('''
+    SELECT DISTINCT ?x
+    WHERE{
+            ?x rdf:type ns:Person.
+            ?x FOAF:knows ns:RockySmith
```
Comment (dgarijo) [inline Assignment4/miguelangel-suarez-22c010/task07.py:91]: The exercise is asking to list the names of the person

File: Assignment4/miguelangel-suarez-22c010/task07.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+  lista += [s]
+print(lista_def)
+
+"""With SPARQL"""
+q1 = prepareQuery('''
+    SELECT Distinct ?x
+    WHERE{
+            ?x FOAF:knows ?y
+    }''', initNs = {"rdf": RDF, "ns": ns, "FOAF":FOAF}
+    )
```
Comment (dgarijo) [inline Assignment4/miguelangel-suarez-22c010/task07.py:114]: In both rdflib and sparql, you are listing all those who know 1 person, not 2 as the exercise requests

**Review by dgarijo — CHANGES_REQUESTED** [review]
Please address the issues highlighted in my review

**Comment by miguelangel-suarez** [comment]
Hope you can see the commits I made for changing the assignment. If any problem or resolution is needed, just tell me.

**Review by dgarijo — APPROVED** [review]

**Pull request merged**